lightly/loss/ntx_ent_loss.py

```diff
@@ -136,37 +115,77 @@ def forward(self,
         # out1 requires a gradient, otherwise keep the same vectors in the 
         # memory bank (this allows for keeping the memory bank constant e.g.
         # for evaluating the loss on the test set)
+        # out1: shape: (batch_size, embedding_size)
+        # negatives: shape: (embedding_size, memory_bank_size)
         out1, negatives = \
             super(NTXentLoss, self).forward(out1, update=out0.requires_grad)
 
+        # We use the cosine similarity, which is a dot product (einsum) here,
+        # as all vectors are already normalized to unit length.
+        # Notation in einsum: n = batch_size, c = embedding_size and k = memory_bank_size.
+
+
         if negatives is not None:
-            negatives = negatives.to(device)
+            # sim_pos is of shape (batch_size, 1) and sim_pos[i] denotes the similarity
+            # of the i-th sample in the batch to its positive pair
+            sim_pos = torch.einsum('nc,nc->n', out0, out1).unsqueeze(-1)
+
             # use negatives from memory bank
-            l_pos = torch.einsum('nc,nc->n', [out0, out1]).unsqueeze(-1)
-            l_neg = torch.einsum('nc,ck->nk', [out0, negatives.clone().detach()])
-        else:
-            # use other samples from batch as negatives
-            output = torch.cat((out0, out1), axis=0)
-            similarity_matrix = self.similarity_function(output, output)
+            negatives = negatives.to(device)
 
-            # filter out the scores from the positive samples
-            l_pos = torch.diag(similarity_matrix, batch_size)
-            r_pos = torch.diag(similarity_matrix, -batch_size)
-            l_pos = torch.cat([l_pos, r_pos]).view(2 * batch_size, 1)
+            # sim_neg each is of shape (batch_size, memory_bank_size) and sim_neg[i,j] denotes the similarity
+            # of the i-th sample to the j-th negative sample
+            sim_neg_out0 = torch.einsum('nc,ck->nk', out0, negatives.clone().detach())
+            sim_neg_out1 = torch.einsum('nc,ck->nk', out1, negatives.clone().detach())
+            sim_neg = sim_neg_out0 + sim_neg_out1
```

Comment:
The only problem I see here is that now, it's a symmetrized version of the loss before. Can you add a unit test to make sure that the output is the same for the original implementation of the loss and this one in the case of memory bank? I believe you will need to remove the symmetrization again to make it work... 

Maybe FIY: The implementation of the loss with memory bank comes directly from the MoCo reference implementation :)

Reply:
I agree and changed it back to the original version.